python/ql/src/experimental/Security/CWE-770/UnicodeDoS.ql

```diff
@@ -0,0 +1,121 @@
+ * @description A remote user-controlled data can reach a costly Unicode normalization with either form NFKC or NFKD. On Windows OS, with an attack such as the One Million Unicode Characters, this could lead to a denial of service. And, with the use of special Unicode characters, like U+2100 (℀) or U+2105 (℅), the payload size could be tripled.
+ * @kind path-problem
+ * @id py/unicode-dos
+ * @precision high
+ * @problem.severity error
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-770
+ */
+
+import python
+import semmle.python.ApiGraphs
+import semmle.python.Concepts
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.DataFlowPublic
+import semmle.python.dataflow.new.RemoteFlowSources
+
+// The Unicode compatibility normalization calls from unicodedata, unidecode, pyunormalize
+// and textnorm modules. The use of argIdx is to constraint the argument being normalized.
+class UnicodeCompatibilityNormalize extends API::CallNode {
+  int argIdx;
+
+  UnicodeCompatibilityNormalize() {
+    exists(API::CallNode cn, DataFlow::Node form |
+      cn = API::moduleImport("unicodedata").getMember("normalize").getACall() and
+      form.asExpr().(StrConst).getS() in ["NFKC", "NFKD"] and
+      TaintTracking::localTaint(form, cn.getArg(0)) and
+      this = cn and
+      argIdx = 1
+    )
+    or
+    exists(API::CallNode cn |
+      cn = API::moduleImport("unidecode").getMember("unidecode").getACall() and
+      this = cn and
+      argIdx = 0
+    )
+    or
+    exists(API::CallNode cn |
+      cn = API::moduleImport("pyunormalize").getMember(["NFKC", "NFKD"]).getACall() and
+      this = cn and
+      argIdx = 0
+    )
+    or
+    exists(API::CallNode cn, DataFlow::Node form |
+      cn = API::moduleImport("pyunormalize").getMember("normalize").getACall() and
+      form.asExpr().(StrConst).getS() in ["NFKC", "NFKD"] and
+      TaintTracking::localTaint(form, cn.getArg(0)) and
+      this = cn and
+      argIdx = 1
+    )
+    or
+    exists(API::CallNode cn, DataFlow::Node form |
+      cn = API::moduleImport("textnorm").getMember("normalize_unicode").getACall() and
+      form.asExpr().(StrConst).getS() in ["NFKC", "NFKD"] and
+      TaintTracking::localTaint(form, cn.getArg(1)) and
+      this = cn and
+      argIdx = 0
+    )
+  }
+
+  DataFlow::Node getPathArg() { result = this.getArg(argIdx) }
+}
+
+predicate underAValue(DataFlow::GuardNode g, ControlFlowNode node, boolean branch) {
+  exists(CompareNode cn | cn = g |
+    exists(API::CallNode lenCall, Cmpop op_gt, Cmpop op_lt, Node n |
+      lenCall = n.getALocalSource() and
+      (
+        // arg <= LIMIT OR arg < LIMIT
+        (op_lt = any(LtE lte) or op_lt = any(Lt lt)) and
+        branch = true and
+        cn.operands(n.asCfgNode(), op_lt, _)
+        or
+        // LIMIT >= arg OR LIMIT > arg 
+        (op_gt = any(GtE gte) or op_gt = any(Gt gt)) and
+        branch = true and
+        cn.operands(_, op_gt, n.asCfgNode())
```

Comment:
If you add the two remaining cases (where `branch = false`) you should also be able to handle code such as
```python
if not len(r) > 1_000:
```

Reply:
added more tests to handle this case https://github.com/github/codeql/pull/15319/commits/dfe6d09c71744273e17e8ccf80f9321b607806d6

Reply:
in ql comparison code, I considered two more cases `if not len(r) > 1_000:` and `if not 1_000 < len(r):`.

Reply:
The expected file was updated accordingly.